File: tests/refs/rabin_karp_search.cpp
// Reference C++ driver for tests/cases/rabin_karp_search.json.
#include <iostream>
#include <sstream>
#include <string>
#include <vector>

std::vector<int> rabinKarpSearch(const std::string& t, const std::string& p) {
    const long long base = 256, prime = 1000000007LL;
    int n = (int)t.size(), m = (int)p.size();
    std::vector<int> out;
    if (m == 0 || m > n) return out;
    long long h = 1;
    for (int i = 0; i < m - 1; ++i) h = (h * base) % prime;
    long long ph = 0, th = 0;
    for (int i = 0; i < m; ++i) {
        ph = (base * ph + (unsigned char)p[i]) % prime;
        th = (base * th + (unsigned char)t[i]) % prime;
    }
    for (int i = 0; i <= n - m; ++i) {
        if (ph == th && t.compare(i, m, p) == 0) out.push_back(i);
        if (i < n - m) {
            th = (base * (th - (unsigned char)t[i] * h) + (unsigned char)t[i + m]) % prime;
            if (th < 0) th += prime;
        }
    }
    return out;
}

static std::string parseStr(const std::string& line) {
    std::size_t sp = line.find(' ', 4);
    if (sp == std::string::npos) return std::string();
    int len = std::stoi(line.substr(4, sp - 4));
    std::string s = line.substr(sp + 1);
    if ((int)s.size() < len) s += std::string(len - s.size(), ' ');
    else if ((int)s.size() > len) s = s.substr(0, len);
    return s;
}

int main() {
    std::string line, name, t, pat; std::vector<int> expected;
    int strIdx = 0; bool he = false; int p = 0, f = 0;
    auto run = [&]() {
        if (name.empty() || strIdx < 2 || !he) return;
        auto got = rabinKarpSearch(t, pat);
        if (got == expected) { std::cout << "PASS rabin_karp_search :: " << name << "\n"; ++p; }
        else { std::cout << "FAIL rabin_karp_search :: " << name << "\n"; ++f; }
        name.clear(); t.clear(); pat.clear(); expected.clear(); strIdx = 0; he = false;
    };
    while (std::getline(std::cin, line)) {
        if (line.empty()) { run(); continue; }
        if (line.rfind("CASE ", 0) == 0) { name = line.substr(5); strIdx = 0; }
        else if (line.rfind("STR ", 0) == 0) {
            std::string s = parseStr(line);
            if (strIdx == 0) { t = s; ++strIdx; } else { pat = s; ++strIdx; }
        } else if (line.rfind("ARR ", 0) == 0) {
            std::istringstream iss(line); std::string tag; iss >> tag;
            std::size_t n; iss >> n;
            expected.clear();
            for (std::size_t i = 0; i < n; ++i) { int v; iss >> v; expected.push_back(v); }
            he = true;
        }
    }
    run();
    std::cout << "TOTAL: " << p << " passed, " << f << " failed\n";
    return f == 0 ? 0 : 1;
}

File: tests/refs/rabin_karp_search.rs
// Reference Rust driver for tests/cases/rabin_karp_search.json.
use std::io::{self, BufRead};

fn rabin_karp(t: &[u8], p: &[u8]) -> Vec<i32> {
    let base: i64 = 256;
    let prime: i64 = 1_000_000_007;
    let (n, m) = (t.len(), p.len());
    let mut out = Vec::new();
    if m == 0 || m > n { return out; }
    let mut h: i64 = 1;
    for _ in 0..m-1 { h = (h * base) % prime; }
    let mut ph: i64 = 0; let mut th: i64 = 0;
    for i in 0..m {
        ph = (base * ph + p[i] as i64) % prime;
        th = (base * th + t[i] as i64) % prime;
    }
    for i in 0..=n-m {
        if ph == th && &t[i..i+m] == p { out.push(i as i32); }
        if i < n - m {
            th = (base * (th - (t[i] as i64) * h) + (t[i+m] as i64)) % prime;
            if th < 0 { th += prime; }
        }
    }
    out
}

fn parse_str(line: &str) -> String {
    let rest = &line[4..];
    if let Some(sp) = rest.find(' ') {
        let len: usize = rest[..sp].parse().unwrap_or(0);
        let s: String = rest[sp+1..].to_string();
        if s.chars().count() < len { let mut s = s; s.push_str(&" ".repeat(len - s.chars().count())); s }
        else { s }
    } else { String::new() }
}

fn main() {
    let stdin = io::stdin();
    let mut name = String::new();
    let mut t = String::new(); let mut pat = String::new();
    let mut expected: Vec<i32> = Vec::new();
    let mut str_idx = 0i32; let mut he = false;
    let mut p = 0u32; let mut f = 0u32;
    let flush = |name: &mut String, t: &mut String, pat: &mut String, expected: &mut Vec<i32>,
                 str_idx: &mut i32, he: &mut bool, p: &mut u32, f: &mut u32| {
        if !name.is_empty() && *str_idx >= 2 && *he {
            let got = rabin_karp(t.as_bytes(), pat.as_bytes());
            if got == *expected { println!("PASS rabin_karp_search :: {}", name); *p += 1; }
            else { println!("FAIL rabin_karp_search :: {}\n  expected: {:?}\n  got: {:?}", name, expected, got); *f += 1; }
        }
        name.clear(); t.clear(); pat.clear(); expected.clear(); *str_idx = 0; *he = false;
    };
    for line in stdin.lock().lines() {
        let line = line.unwrap_or_default();
        if line.is_empty() { flush(&mut name, &mut t, &mut pat, &mut expected, &mut str_idx, &mut he, &mut p, &mut f); continue; }
        if let Some(rest) = line.strip_prefix("CASE ") { name = rest.to_string(); str_idx = 0; }
        else if line.starts_with("STR ") {
            let s = parse_str(&line);
            if str_idx == 0 { t = s; str_idx += 1; } else { pat = s; str_idx += 1; }
        }
        else if line.starts_with("ARR ") {
            let mut parts = line.split_whitespace(); parts.next();
            let n: usize = parts.next().and_then(|s| s.parse().ok()).unwrap_or(0);
            expected.clear();
            for _ in 0..n { if let Some(t) = parts.next() { if let Ok(v) = t.parse::<i32>() { expected.push(v); } } }
            he = true;
        }
    }
    flush(&mut name, &mut t, &mut pat, &mut expected, &mut str_idx, &mut he, &mut p, &mut f);
    println!("TOTAL: {} passed, {} failed", p, f);
    std::process::exit(if f == 0 { 0 } else { 1 });
}
